lib/zk_recipes/cache.rb

```diff
@@ -35,39 +49,149 @@ def initialize(logger: nil, host: nil, timeout: nil, zk_opts: {})
+      raise Error, "path already registered path=#{path.inspect}" if @registered_paths.key?(path)
 
-      @cache[path] = CachedPath.new(default_value)
-      @registered_values[path] = RegisteredPath.new(default_value, block)
+      @registered_paths[path] = RegisteredPath.new(default_value, deserializer)
       @logger&.debug { "added path=#{path} default_value=#{default_value.inspect}" }
       ActiveSupport::Notifications.instrument(AS_NOTIFICATION, path: path, value: default_value)
     end
 
+    # Can only be called before setup_callbacks
+    # This method is not threadsafe. All configuration should be done on a single thread.
+    def register_children(path, default_children = [], &deserializer)
+      raise Error, "register_children only allowed before setup_callbacks called" if @zk
+      raise Error, "path already registered path=#{path.inspect}" if @registered_children_paths.key?(path)
+
+      @registered_children_paths[path] = RegisteredPath.new(default_children, deserializer)
+      @logger&.debug { "added children=#{path} default_children=#{default_children.inspect}" }
+      ActiveSupport::Notifications.instrument(AS_NOTIFICATION_CHILDREN, path: path, value: default_children)
+    end
+
+    # Can only be called after setup_callbacks
+    #
+    # This method is threadsafe.
+    # All accesses in this method need to be race-free.
+    def register_runtime(path, default_value, &deserializer)
+      # This check is threadsafe since @zk is only set once, (in `#setup_callbacks`). It is never changed after that.
+      raise Error, "register_runtime only allowed after setup_callbacks called" unless @zk
+
+      # This is not inherently threadsafe, but since you can't register non-runtime paths after `#setup_callbacks` is
+      # called, and `#setup_callbacks` sets @zk, we know that `#setup_callbacks` has run, @zk is set, and no new paths
+      # can be registered.
+      raise RuntimeRegistrationError, "already registered path=#{path.inspect}" if registered_path?(path)
+
+      already_registered = true
+
+      # The registered_runtime_path needs to be set before the watch is registered, but they have two different locks.
+      # Use the registered_runtime_watch as the authoritative lock so it can be held, set the registered_runtime_path,
+      # then set the registered_runtime_watch. This means that to check if a runtime path is set, you must check if the
+      # watch is set, not if the path is set.
+      @registered_runtime_watches.compute_if_absent(path) do
+        already_registered = false
+
+        # The watch wasn't set, so the runtime path should *not* exist. If it does, it means there is a bug.
+        if @registered_runtime_paths.put_if_absent(path, RegisteredPath.new(default_value, deserializer))
+          @registered_runtime_paths.delete(path)
+          raise StateError, "runtime path=#{path.inspect} should not exist if the watch is absent"
```

Comment:
Should this `raise StateError`, or just log the error and overwrite the path? Thinking it should log and overwrite. Tests already fail if there are any unexpected errors logged.

Reply:
Preference towards self-recovery over raising exceptions where possible

Reply:
👍